### PR TITLE
Update the implementation of ClassOrdering to handle false negatives

### LINKED ISSUE #3809: False negative for `ClassOrdering`
## Expected Behavior
For `ClassOrdering`, code like the following should report violations.
```
class MultipleMisorders(private val x: String) {
    companion object {
        const val IMPORTANT_VALUE = 3
    }
    
    class Nested { }
    
    fun returnX() = x
    
    class Nested2 { }
    
    constructor(z: Int): this(z.toString())
    
    class Nested3 { }
    
    val y = x
}
```

## Observed Behavior
There are no violations reported.

## Steps to Reproduce
Use the provided code snippet.

## Context
There were many violations detected in my company's projects when upgraded to 1.17, however, there are still false negatives remaining in the codebase.

## Your Environment
* Version of detekt used: 1.17

### PR DESCRIPTION
**Pull request by chao2zhang**
Resolves #3809.

With this new implementation, one noticeable changes are that the number of violations is going to be increased:
For example, in a class layout like

```
class MyClass {
    val a: Int
    fun b(): String
    val c: Int
    val d: Int
    val e: Int
    val f: Int
}
```
Previous the violation will be reported only the boundary (a <-> b, b <-> c), now it will be reported on (c, d, e, f).
If the increase in the violations is a concern, we can implement grouping logic to reduce the number of violations.

### CODE REVIEW

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ClassOrdering.kt [hunk 1]
```diff
@@ -61,59 +61,65 @@ class ClassOrdering(config: Config = Config.empty) : Rule(config) {
...
+    this is KtProperty -> Section(0)
+    this is KtClassInitializer -> Section(0)
+    this is KtSecondaryConstructor -> Section(1)
+    this is KtNamedFunction -> Section(2)
+    this is KtObjectDeclaration && isCompanion() -> Section(3)
+    else -> null
+}
+
+@JvmInline
+@Suppress("MagicNumber", "ModifierOrder") // TODO Remove ModifierOrder once value class is supported.
```
Comment (chao2zhang) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ClassOrdering.kt:109]: This is pending #3719

**Comment by BraisGabin** [comment]
Time ago I did some research regarging this rule and the number of issue that it raises. I found that we could use ["longest increasing subsequence (LIS)"](https://www.geeksforgeeks.org/minimum-insertions-sort-array/) to reduce the number of issues. But I never implement it. I don't know even if we need to implement it. But as you talk about it I thought it could be usefull.

**Review by BraisGabin — APPROVED** [review]

**Review by cortinico — APPROVED** [review]

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ClassOrdering.kt [hunk 2]
```diff
@@ -61,59 +61,65 @@ class ClassOrdering(config: Config = Config.empty) : Rule(config) {
...
+                        CodeSmell(
+                            issue = issue,
+                            entity = Entity.from(ktDeclaration),
+                            message = message,
+                            references = listOf(Entity.from(classBody))
+                        )
                     )
                 }
-            )
+                section != null && section > currentSection -> currentSection = section
```
Comment (schalkms) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ClassOrdering.kt:83]: Can we move out the `section != null` check in order to not check it twice?

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ClassOrdering.kt [hunk 3]
```diff
@@ -61,59 +61,65 @@ class ClassOrdering(config: Config = Config.empty) : Rule(config) {
...
+}
+
+@Suppress("MagicNumber")
+private fun KtDeclaration.toSection(): Section? = when {
+    this is KtProperty -> Section(0)
+    this is KtClassInitializer -> Section(0)
+    this is KtSecondaryConstructor -> Section(1)
+    this is KtNamedFunction -> Section(2)
+    this is KtObjectDeclaration && isCompanion() -> Section(3)
+    else -> null
```
Comment (schalkms) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ClassOrdering.kt:105]: Is there any alternative to the `null` return? Then we could spare the null checks above?

**Review by schalkms — COMMENTED** [review]
Thanks for the fix!

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ClassOrdering.kt (hunk 3 above)
Comment (chao2zhang) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ClassOrdering.kt:105]: Kotlin documentation does not specify where we should put nested classes or objects, so the ordering for nested classes shouldn't matter from a static analysis point of view, hence it should not have any priority order - `null` is a reasonable choice.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ClassOrdering.kt (hunk 2 above)
Comment (chao2zhang) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ClassOrdering.kt]: Refactored using `continue` in a `for loop`.

**Pull request merged**